Firestore/core/src/firebase/firestore/core/query.h

```diff
@@ -239,11 +245,12 @@ class Query {
   // The memoized list of sort orders.
   mutable OrderByList memoized_order_bys_;
 
-  int32_t limit_ = kNoLimit;
+  int32_t limit_ = Target::kNoLimit;
   std::shared_ptr<Bound> start_at_;
   std::shared_ptr<Bound> end_at_;
 
-  mutable std::string canonical_id_;
+  // The corresponding Target of this Query instance.
+  mutable std::shared_ptr<const Target> memoized_target;
```

Comment:
Given that our only known divergence between Query and Target is going to be an additional property, it seems wasteful to duplicate the query properties (the path, filter list, etc) in both of these classes.

You could avoid having to keep this separate pointer by making Target a direct member and fully delegating all common properties to it.